// vue3-marquee/dist/vue3-marquee.ssr.js
'use strict';var vue=require('vue');function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];

  if (_i == null) return;
  var _arr = [];
  var _n = true;
  var _d = false;

  var _s, _e;

  try {
    for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}var script = /*#__PURE__*/vue.defineComponent({
  name: "Vue3Marquee",
  props: {
    direction: {
      type: String,
      required: false,
      default: "left",
      validator: function validator(value) {
        return ["left", "right"].includes(value);
      }
    },
    duration: {
      type: Number,
      required: false,
      default: 20
    },
    delay: {
      type: Number,
      required: false,
      default: 0
    },
    loop: {
      type: Number,
      required: false,
      default: 0
    },
    clone: {
      type: Boolean,
      required: false,
      default: false
    },
    gradient: {
      type: Boolean,
      required: false,
      default: false
    },
    gradientColor: {
      type: Array,
      required: false,
      default: [255, 255, 255],
      validator: function validator(value) {
        if (value.length != 3) {
          return false;
        }

        if (typeof value[0] !== "number" || value[0] < 0 || value[0] > 255) {
          return false;
        }

        if (typeof value[1] !== "number" || value[1] < 0 || value[1] > 255) {
          return false;
        }

        if (typeof value[2] !== "number" || value[2] < 0 || value[2] > 255) {
          return false;
        }

        return true;
      }
    },
    gradientWidth: {
      type: [String, Number],
      required: false,
      default: 200
    },
    pauseOnHover: {
      type: Boolean,
      required: false,
      default: false
    },
    pauseOnClick: {
      type: Boolean,
      required: false,
      default: false
    },
    options: {
      type: Object,
      required: false,
      default: {}
    }
  },
  data: function data() {
    return {
      localDirection: this.direction,
      localDuration: this.duration,
      localDelay: this.delay,
      localLoop: this.loop,
      localGradient: this.gradient,
      localGradientColor: this.gradientColor,
      localGradientWidth: this.gradientWidth,
      localPauseOnHover: this.pauseOnHover,
      localPauseOnClick: this.pauseOnClick,
      localClone: this.clone,
      minWidth: 0,
      cloneAmount: 0
    };
  },
  methods: {
    checkForClone: function checkForClone() {
      if (this.localClone) {
        this.minWidth = 0;
        var contentWidth = this.$refs.marqueeContent.clientWidth;
        var containerWidth = this.$refs.marqueeContainer.clientWidth;
        this.cloneAmount = Math.ceil(containerWidth / (contentWidth * 2));
        console.log(this.cloneAmount);
      } else {
        this.minWidth = "100%";
      }
    }
  },
  computed: {
    getMarqueeDirection: function getMarqueeDirection() {
      if (this.localDirection === "left") {
        return "normal";
      } else if (this.localDirection === "right") {
        return "reverse";
      }

      return "normal";
    },
    getPauseOnHover: function getPauseOnHover() {
      if (this.localPauseOnHover) {
        return "paused";
      }

      return "running";
    },
    getpauseOnClick: function getpauseOnClick() {
      if (this.localPauseOnClick) {
        return "paused";
      }

      return "running";
    },
    getLoops: function getLoops() {
      return this.localLoop === 0 ? "infinite" : this.localLoop;
    },
    getCurrentStyle: function getCurrentStyle() {
      var cssVariables = {
        "--duration": "".concat(this.localDuration, "s"),
        "--delay": "".concat(this.localDelay, "s"),
        "--direction": "".concat(this.getMarqueeDirection),
        "--pauseOnHover": "".concat(this.getPauseOnHover),
        "--pauseOnClick": "".concat(this.getpauseOnClick),
        "--loops": "".concat(this.getLoops)
      };

      if (this.localGradient) {
        var rgbaGradientColor = "rgba(".concat(this.localGradientColor[0], ", ").concat(this.localGradientColor[1], ", ").concat(this.localGradientColor[2]);
        cssVariables["--gradient-color"] = "".concat(rgbaGradientColor, ", 1), ").concat(rgbaGradientColor, ", 0)");
        cssVariables["--gradient-width"] = typeof this.localGradientWidth === "number" ? "".concat(this.localGradientWidth, "px") : this.localGradientWidth;
      }

      cssVariables["--min-width"] = this.minWidth;
      return cssVariables;
    }
  },
  mounted: function mounted() {
    if (this.options) {
      if (this.options.direction) {
        this.localDirection = this.options.direction;
      }

      if (this.options.duration) {
        this.localDuration = this.options.duration;
      }

      if (this.options.delay) {
        this.localDelay = this.options.delay;
      }

      if (this.options.loop) {
        this.localLoop = this.options.loop;
      }

      if (this.options.gradient) {
        this.localGradient = this.options.gradient;
      }

      if (this.options.gradientColor) {
        this.localGradientColor = this.options.gradientColor;
      }

      if (this.options.gradientWidth) {
        this.localGradientWidth = this.options.gradientWidth;
      }

      if (this.options.pauseOnHover) {
        this.localPauseOnHover = this.options.pauseOnHover;
      }

      if (this.options.pauseOnClick) {
        this.localPauseOnClick = this.options.pauseOnClick;
      }

      if (this.options.clone) {
        this.localClone = this.options.clone;
      }
    }

    this.checkForClone();
  }
});var _hoisted_1 = {
  class: "overlay",
  ref: "marqueeContainer"
};
var _hoisted_2 = {
  class: "marquee",
  ref: "marqueeContent"
};
var _hoisted_3 = {
  class: "marquee"
};
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("div", null, [vue.createElementVNode("div", {
    class: "vue3-marquee",
    style: vue.normalizeStyle(_ctx.getCurrentStyle)
  }, [vue.createElementVNode("div", _hoisted_1, null, 512), vue.createElementVNode("div", _hoisted_2, [vue.renderSlot(_ctx.$slots, "default")], 512), vue.createElementVNode("div", _hoisted_3, [vue.renderSlot(_ctx.$slots, "default")]), (vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(_ctx.cloneAmount, function (num) {
    return vue.withDirectives((vue.openBlock(), vue.createElementBlock("div", {
      class: "marquee",
      key: num
    }, [vue.renderSlot(_ctx.$slots, "default")], 512)), [[vue.vShow, _ctx.localClone]]);
  }), 128))], 4)]);
}function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}var css_248z = "\n.vue3-marquee {\n  overflow-x: hidden !important;\n  display: flex !important;\n  flex-direction: row !important;\n  position: relative;\n  width: 100%;\n}\n.vue3-marquee:hover div {\n  animation-play-state: var(--pauseOnHover);\n}\n.vue3-marquee:active div {\n  animation-play-state: var(--pauseOnClick);\n}\n.marquee {\n  flex: 0 0 auto;\n  min-width: var(--min-width);\n  z-index: 1;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  animation: scroll var(--duration) linear var(--delay) var(--loops);\n  animation-direction: var(--direction);\n}\n@keyframes scroll {\n0% {\n    transform: translateX(0%);\n}\n100% {\n    transform: translateX(-100%);\n}\n}\n.overlay {\n  position: absolute;\n  width: 100%;\n  height: 100%;\n}\n.overlay::before,\n.overlay::after {\n  background: linear-gradient(to right, var(--gradient-color));\n  content: \"\";\n  height: 100%;\n  position: absolute;\n  width: var(--gradient-width);\n  z-index: 2;\n}\n.overlay::after {\n  right: 0;\n  top: 0;\n  transform: rotateZ(180deg);\n}\n.overlay::before {\n  left: 0;\n  top: 0;\n}\n";
styleInject(css_248z);script.render = render;// Import vue component
// IIFE injects install function into component, allowing component
// to be registered via Vue.use() as well as Vue.component(),

var component = /*#__PURE__*/(function () {
  // Get component instance
  var installable = script; // Attach install function executed by Vue.use()

  installable.install = function (app) {
    app.component('Vue3Marquee', installable);
  };

  return installable;
})(); // It's possible to expose named exports when writing components that can
// also be used as directives, etc. - eg. import { RollupDemoDirective } from 'rollup-demo';
// export const RollupDemoDirective = directive;
var namedExports=/*#__PURE__*/Object.freeze({__proto__:null,'default':component});// only expose one global var, with named exports exposed as properties of
// that global var (eg. plugin.namedExport)

Object.entries(namedExports).forEach(function (_ref) {
  var _ref2 = _slicedToArray(_ref, 2),
      exportName = _ref2[0],
      exported = _ref2[1];

  if (exportName !== 'default') component[exportName] = exported;
});module.exports=component;
